File: PersonalNotes/BaseDBHelper.cs
using System;
using System.Collections.Generic;
using System.Data.SQLite;
using Dapper;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

/// Классы для работы с базой данных
/// Создание и открытие соединения, получения информации об ошибках
/// добавления, изменения и удаления записей в записной книжке

namespace PersonalNotes
{
    /// <summary>
    /// Базовый класс для работы с базой данных
    /// создание и открытие соединения
    /// проверка открытия базы данных, закрытие соединения
    /// </summary>
    class BaseDBHelper : IDisposable
    {
        protected SQLiteConnection m_conn; // объект соединееия
        protected string m_errorText; // строка с ошибкой
        // строка ошибки
        public string errorText { get { return m_errorText; } }
        // открыто ли  соединение
        public bool isOpen { get { return m_conn.State == System.Data.ConnectionState.Open; } }

        /// <summary>
        /// Создать объект соединения и открыть созаное соединение
        /// </summary>
        /// <param name="dataFileName"></param>
        public BaseDBHelper(string dataFileName)
        {
            m_errorText = string.Empty;
            try
            {
                SQLiteFactory factory = new SQLiteFactory();
                m_conn = (SQLiteConnection)factory.CreateConnection();
                m_conn.ConnectionString = String.Format("Data Source={0}; Version=3;", dataFileName);
                m_conn.Open();
            }
            catch (Exception ex)
            {
                m_errorText = ex.Message;
            }
        }
        /// <summary>
        /// Закрыть соединение
        /// </summary>
        public void Dispose()
        {
            if (isOpen)
                m_conn.Close();
        }
    }
    /// <summary>
    /// Производный класс от BaseDBHelper для управления данными
    /// получение записей, добаление, изменение, удаление данных
    /// </summary>
    class DBHelper : BaseDBHelper
    {
        public DBHelper(string dataFileName) : base(dataFileName)
        {

        }

        /// <summary>
        /// Выдать список адресов
        /// </summary>
        /// <returns>список объектов, представление записей</returns>
        public async Task<List<AddressBook>> GetAddressData(string letter = "")
        {
            List<AddressBook> lst = null;
            string sqlText = string.IsNullOrEmpty(letter) ? "select id, name, birth_date, phone, address, comments from address_book order by name" :
                $"select id, name, birth_date, phone, address, comments from address_book where substr(name,1,1) = '{letter}' order by name";
            try
            {
                var task = await m_conn.QueryAsync<AddressBook>(sqlText);
                lst = task.ToList();
            }
            catch (Exception ex)
            {
                m_errorText = ex.Message;
            }
            return lst;
        }
        /// <summary>
        /// Возвратить запись из адресной книги по идентификатору
        /// </summary>
        /// <param name="idrec"></param>
        /// <returns></returns>
        public AddressBook GetAddressDataRecord(long idrec)
        {
            AddressBook rec = null;
            string sqlText = $"select id, name, birth_date, phone, address, comments from address_book where id ={idrec}";
            try
            {
                rec = m_conn.QueryFirstOrDefault<AddressBook>(sqlText);
            }
            catch (Exception ex)
            {
                m_errorText = ex.Message;
            }

            return rec;
        }
        /// <summary>
        /// Добавить запись в адресную книжку
        /// </summary>
        /// <param name="rec">реквизиты для добавления записи</param>
        /// <returns>число добавленных записей</returns>
        public int AddAddressRecord(AddressBook rec)
        {
            int recs = 0;
            string sdate = rec.birth_date.ToString("yyyy-MM-dd HH:mm:ss");
            string sqlText = $"insert into address_book(name, phone, birth_date, address,comments) values('{rec.name}', '{rec.phone}', '{sdate}', '{rec.address}', '{rec.comments}')";
            try 
            {
                recs = m_conn.Execute(sqlText);
            } 
            catch(Exception ex) 
            { 
                m_errorText = ex.Message;
            }
            
            return recs;
        }
        /// <summary>
        /// Изменить запись в адресной книжке по шаблону
        /// </summary>
        /// <param name="rec">объект, шаблон для изменения записи</param>
        /// <returns></returns>
        public int UpdateAdressRecord(AddressBook rec) 
        {
            int recs = 0;
            string sdate = rec.birth_date.ToString("yyyy-MM-dd HH:mm:ss");
            string sqlText = $"update address_book set name = '{rec.name}', phone = '{rec.phone}', birth_date = '{sdate}', address = '{rec.address}',comments = '{rec.comments}' where id = {rec.id}";
            try
            {
                recs = m_conn.Execute(sqlText);
            }
            catch (Exception ex)
            {
                m_errorText = ex.Message;
            }

            return recs;
        }
        /// <summary>
        /// Удаление записи в адресной книжке
        /// </summary>
        /// <param name="id">идентификатор удаляемой записи</param>
        /// <returns></returns>
        public int DeleteAddressRecord(long id)
        {
            int recs = 0;
            string sqlText = $"delete from address_book where id = {id}";
            try
            {
                recs = m_conn.Execute(sqlText);
            }
            catch (Exception ex)
            {
                m_errorText = ex.Message;
            }

            return recs;

        }
        /// <summary>
        /// Выдать список заметок в записной книжке
        /// </summary>
        /// <returns>список объектов, представлений записей</returns>
        public async Task<List<Note>> GetNotes()
        {
            List<Note> lst = null;
            string sqlText = "select id, note_datetime, comments from notes order by note_datetime";
            try
            {
                var task = await m_conn.QueryAsync<Note>(sqlText);
                lst = task.ToList();
            }
            catch (Exception ex)
            {
                m_errorText = ex.Message;
            }
            return lst;
        }
        /// <summary>
        /// Возвратить запись из заметок по идентификатору
        /// </summary>
        /// <param name="idrec"></param>
        /// <returns></returns>
        public Note GetNotesDataRecord(long idrec)
        {
            Note rec = null;
            string sqlText = $"select id, note_datetime, comments from notes where id ={idrec}";
            try
            {
                rec = m_conn.QueryFirstOrDefault<Note>(sqlText);
            }
            catch (Exception ex)
            {
                m_errorText = ex.Message;
            }

            return rec;
        }

        /// <summary>
        /// Добавить запись в заметки
        /// </summary>
        /// <param name="rec">реквизиты для добавления записи</param>
        /// <returns>число добавленных записей</returns>
        public int AddNoteRecord(Note rec)
        {
            int recs = 0;
            string sdate = rec.note_datetime.ToString("yyyy-MM-dd HH:mm:ss");
            string sqlText = $"insert into notes(note_datetime,comments) values( '{sdate}', '{rec.comments}')";
            try
            {
                recs = m_conn.Execute(sqlText);
            }
            catch (Exception ex)
            {
                m_errorText = ex.Message;
            }

            return recs;
        }
        /// <summary>
        /// Изменить запись по шаблону
        /// </summary>
        /// <param name="rec">объект, шаблон для изменения записи</param>
        /// <returns></returns>
        public int UpdateNotesRecord(Note rec)
        {
            int recs = 0;
            string sdate = rec.note_datetime.ToString("yyyy-MM-dd HH:mm:ss");
            string sqlText = $"update notes set note_datetime = '{sdate}', comments = '{rec.comments}' where id = {rec.id}";
            try
            {
                recs = m_conn.Execute(sqlText);
            }
            catch (Exception ex)
            {
                m_errorText = ex.Message;
            }

            return recs;
        }
        /// <summary>
        /// Удаление записи в записной книжке
        /// </summary>
        /// <param name="id">идентификатор удаляемой записи</param>
        /// <returns></returns>
        public int DeleteNotesRecord(long id)
        {
            int recs = 0;
            string sqlText = $"delete from notes where id = {id}";
            try
            {
                recs = m_conn.Execute(sqlText);
            }
            catch (Exception ex)
            {
                m_errorText = ex.Message;
            }

            return recs;

        }

    }
}

File: PersonalNotesCPP/BaseDBHelper.cpp
#include "pch.h"
#include "BaseDBHelper.h"

using namespace CppCLRWinFormsProject;

/// <summary>
/// Создать объект соединения и открыть созаное соединение
/// </summary>
/// <param name="dataFileName"></param>
BaseDBHelper::BaseDBHelper(String^ dataFileName)
{
    m_errorText = String::Empty;
    try
    {
        SQLiteFactory^ factory = gcnew SQLiteFactory;
        m_conn = (SQLiteConnection^)factory->CreateConnection();
        m_conn->ConnectionString = String::Format(L"Data Source={0}; Version=3;", dataFileName);
        m_conn->Open();
    }
    catch (Exception^ ex)
    {
        m_errorText = ex->Message;
    }
}
/// <summary>
///  Закрыть открытое соединение
/// </summary>
BaseDBHelper::~BaseDBHelper()
{
    if (isOpen()) m_conn->Close();
}

List<AddressBook^>^ DBHelper::GetAddressData(String^ letter)
{
    List<AddressBook^>^ lst = nullptr;
    String^ sqlText = String::IsNullOrEmpty(letter) ? L"select id, name, birth_date, phone, address, comments from address_book order by name" :
        String::Format(L"select id, name, birth_date, phone, address, comments from address_book where substr(name,1,1) = '{0}' order by name", letter);
    try
    {
        SQLiteCommand^ cmd = m_conn->CreateCommand();
        cmd->CommandText = sqlText;
        SQLiteDataReader^ reader = cmd->ExecuteReader();
        lst = gcnew List<AddressBook^>();
        while (reader->Read())
        {
            AddressBook^ bk = gcnew AddressBook;
            bk->id = Convert::ToInt64(reader["id"]); 
            bk->name = reader["name"]->ToString();
            bk->birth_date = Convert::ToDateTime(reader["birth_date"]);
            bk->phone = reader["phone"]->ToString();
            bk->address = reader["address"]->ToString();
            bk->comments = reader["comments"]->ToString();
            lst->Add(bk);
        }
        reader->Close();
        
    }       
    catch (Exception^ ex)
    {
        m_errorText = ex->Message;
    }
    return lst;

}
/// <summary>
/// Возвратить запись из адресной книги по идентификатору
/// </summary>
/// <param name="idrec"></param>
/// <returns></returns>
AddressBook^ DBHelper::GetAddressDataRecord(long idrec)
{
    AddressBook^ bk = nullptr;
    String^ sqlText = String::Format("select id, name, birth_date, phone, address, comments from address_book where id ={0}", idrec);
    try
    {
        SQLiteCommand^ cmd = m_conn->CreateCommand();
        cmd->CommandText = sqlText;
        SQLiteDataReader^ reader = cmd->ExecuteReader();
        while (reader->Read())
        {
            bk = gcnew AddressBook;
            bk->id = Convert::ToInt64(reader["id"]);
            bk->name = reader["name"]->ToString();
            bk->birth_date = Convert::ToDateTime(reader["birth_date"]);
            bk->phone = reader["phone"]->ToString();
            bk->address = reader["address"]->ToString();
            bk->comments = reader["comments"]->ToString();
        }
        reader->Close();

    }
    catch (Exception^ ex)
    {
        m_errorText = ex->Message;
    }

    return bk;
}

/// <summary>
/// Добавить запись в адресную книжку
/// </summary>
/// <param name="rec">реквизиты для добавления записи</param>
/// <returns>число добавленных записей</returns>
int DBHelper::AddAddressRecord(AddressBook^ rec)
{
    int recs = 0;
    String^ sdate = rec->birth_date->ToString("yyyy-MM-dd HH:mm:ss");
    String^ sqlText = String::Format("insert into address_book(name, phone, birth_date, address,comments) values('{0}', '{1}', '{2}', '{3}', '{4}')",
        rec->name, rec->phone, sdate, rec->address, rec->comments);
    try
    {
        SQLiteCommand^ cmd = m_conn->CreateCommand();
        cmd->CommandText = sqlText;
        recs = cmd->ExecuteNonQuery();
    }
    catch (Exception^ ex)
    {
        m_errorText = ex->Message;
    }

    return recs;
}
/// <summary>
/// Изменить запись в адресной книжке по шаблону
/// </summary>
/// <param name="rec">объект, шаблон для изменения записи</param>
/// <returns></returns>
int DBHelper::UpdateAdressRecord(AddressBook^ rec)
{
    int recs = 0;
    String^ sdate = rec->birth_date->ToString("yyyy-MM-dd HH:mm:ss");
    String^ sqlText = String::Format("update address_book set name = '{0}', phone = '{1}', birth_date = '{2}', address = '{3}',comments = '{4}' where id = {5}",
        rec->name, rec->phone, sdate,rec->address, rec->comments, rec->id);
    try
    {
        SQLiteCommand^ cmd = m_conn->CreateCommand();
        cmd->CommandText = sqlText;
        recs = cmd->ExecuteNonQuery();
    }
    catch (Exception^ ex)
    {
        m_errorText = ex->Message;
    }

    return recs;
}
/// <summary>
/// Удаление записи в адресной книжке
/// </summary>
/// <param name="id">идентификатор удаляемой записи</param>
/// <returns></returns>
int DBHelper::DeleteAddressRecord(long id)
{
    int recs = 0;
    String^ sqlText = String::Format("delete from address_book where id = {0}",id);
    try
    {
        SQLiteCommand^ cmd = m_conn->CreateCommand();
        cmd->CommandText = sqlText;
        recs = cmd->ExecuteNonQuery();
    }
    catch (Exception^ ex)
    {
        m_errorText = ex->Message;
    }

    return recs;

}
/// <summary>
/// Выдать список заметок в записной книжке
/// </summary>
/// <returns>список объектов, представлений записей</returns>
List<Note^>^ DBHelper::GetNotes()
{
    List<Note^>^ lst = nullptr;
    String^ sqlText = "select id, note_datetime, comments from notes order by note_datetime";
    try
    {
        SQLiteCommand^ cmd = m_conn->CreateCommand();
        cmd->CommandText = sqlText;
        SQLiteDataReader^ reader = cmd->ExecuteReader();
        lst = gcnew List<Note^>;
        while (reader->Read())
        {
            Note^ note = gcnew Note;
            note->id = Convert::ToInt64(reader["id"]);
            note->note_datetime = Convert::ToDateTime(reader["note_datetime"]);
            note->comments = reader["comments"]->ToString();
            lst->Add(note);
        }
        reader->Close();
    }
    catch (Exception^ ex)
    {
        m_errorText = ex->Message;
    }
    return lst;
}

/// <summary>
/// Возвратить запись из заметок по идентификатору
/// </summary>
/// <param name="idrec"></param>
/// <returns></returns>
Note^ DBHelper::GetNotesDataRecord(long idrec)
{
    Note^ note = nullptr;
    String^ sqlText = String::Format("select id, note_datetime, comments from notes where id ={0}",idrec);
    try
    {
        SQLiteCommand^ cmd = m_conn->CreateCommand();
        cmd->CommandText = sqlText;
        SQLiteDataReader^ reader = cmd->ExecuteReader();
        while (reader->Read())
        {
            note = gcnew Note;
            note->id = Convert::ToInt64(reader["id"]);
            note->note_datetime = Convert::ToDateTime(reader["note_datetime"]);
            note->comments = reader["comments"]->ToString();
        }
        reader->Close();
    }
    catch (Exception^ ex)
    {
        m_errorText = ex->Message;
    }

    return note;
}

/// <summary>
/// Добавить запись в заметки
/// </summary>
/// <param name="rec">реквизиты для добавления записи</param>
/// <returns>число добавленных записей</returns>
int DBHelper::AddNoteRecord(Note^ rec)
{
    int recs = 0;
    String^ sdate = rec->note_datetime->ToString("yyyy-MM-dd HH:mm:ss");
    String^ sqlText = String::Format("insert into notes(note_datetime,comments) values( '{0}', '{1}')", sdate, rec->comments);
    try
    {
        SQLiteCommand^ cmd = m_conn->CreateCommand();
        cmd->CommandText = sqlText;
        recs = cmd->ExecuteNonQuery();
    }
    catch (Exception^ ex)
    {
        m_errorText = ex->Message;
    }

    return recs;
}
/// <summary>
/// Изменить запись по шаблону
/// </summary>
/// <param name="rec">объект, шаблон для изменения записи</param>
/// <returns></returns>
int DBHelper::UpdateNotesRecord(Note^ rec)
{
    int recs = 0;
    String^ sdate = rec->note_datetime->ToString("yyyy-MM-dd HH:mm:ss");
    String^ sqlText = String::Format("update notes set note_datetime = '{0}', comments = '{1}' where id = {2}", sdate, rec->comments, rec->id);
    try
    {
        SQLiteCommand^ cmd = m_conn->CreateCommand();
        cmd->CommandText = sqlText;
        recs = cmd->ExecuteNonQuery();
    }
    catch (Exception^ ex)
    {
        m_errorText = ex->Message;
    }

    return recs;
}
/// <summary>
/// Удаление записи в записной книжке
/// </summary>
/// <param name="id">идентификатор удаляемой записи</param>
/// <returns></returns>
int DBHelper::DeleteNotesRecord(long id)
{
    int recs = 0;
    String^ sqlText = String::Format("delete from notes where id = {0}",id);
    try
    {
        SQLiteCommand^ cmd = m_conn->CreateCommand();
        cmd->CommandText = sqlText;
        recs = cmd->ExecuteNonQuery();
    }
    catch (Exception^ ex)
    {
        m_errorText = ex->Message;
    }

    return recs;

}



